File: onlinepayments-sdk/src/main/java/com/onlinepayments/sdk/client/android/session/SessionEncryptionHelper.java
/*
 * Copyright 2020 Global Collect Services B.V
 */

package com.onlinepayments.sdk.client.android.session;

import java.security.InvalidParameterException;
import java.util.Map;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.onlinepayments.sdk.client.android.asynctask.EncryptDataAsyncTask;
import com.onlinepayments.sdk.client.android.asynctask.PublicKeyAsyncTask;
import com.onlinepayments.sdk.client.android.Util;
import com.onlinepayments.sdk.client.android.exception.EncryptDataException;
import com.onlinepayments.sdk.client.android.listener.PaymentRequestPreparedListener;
import com.onlinepayments.sdk.client.android.model.PaymentRequest;
import com.onlinepayments.sdk.client.android.model.PreparedPaymentRequest;
import com.onlinepayments.sdk.client.android.model.PublicKeyResponse;
import com.onlinepayments.sdk.client.android.model.api.ErrorResponse;

/**
 * Session contains all methods needed for making a payment.
 *
 * @deprecated In a future release, this class will become internal to the SDK. Use {@link Session#preparePaymentRequest(PaymentRequest, Context, PaymentRequestPreparedListener)} to encrypt your Payment Requests.
 */
@Deprecated
public class SessionEncryptionHelper implements EncryptDataAsyncTask.OnEncryptDataCompleteListener, PublicKeyAsyncTask.OnPublicKeyLoadedListener, PublicKeyAsyncTask.PublicKeyListener {

	private Context context;
	private OnPaymentRequestPreparedListener listener;
	private PaymentRequestPreparedListener paymentRequestPreparedListener;
	private PaymentRequest paymentRequest;
	private String clientSessionId;
	private Map<String, String> metaData;

	/**
	 * Helper for encrypting the payment request.
	 *
	 * @param context used for reading device metadata which is sent to the Online Payments gateway
	 * @param paymentRequest the {@link PaymentRequest} that will be encrypted
	 * @param clientSessionId the sessionId that is used to communicate with the Online Payments gateway
	 * @param listener {@link OnPaymentRequestPreparedListener} that will be called when encryption is completed
	 *
	 *
	 * @deprecated use {@link #SessionEncryptionHelper(PaymentRequest, String, Map, PaymentRequestPreparedListener)} instead.
     */
	@Deprecated
	public SessionEncryptionHelper(Context context, PaymentRequest paymentRequest, String clientSessionId, OnPaymentRequestPreparedListener listener) {

		if (paymentRequest == null ) {
			throw new InvalidParameterException("Error creating SessionEncryptionHelper, paymentRequest may not be null");
		}
		if (context == null ) {
			throw new InvalidParameterException("Error creating SessionEncryptionHelper, context may not be null");
		}
		if (clientSessionId == null ) {
			throw new InvalidParameterException("Error creating SessionEncryptionHelper, clientSessionId may not be null");
		}
		if (listener == null ) {
			throw new InvalidParameterException("Error creating SessionEncryptionHelper, listener may not be null");
		}

		this.context = context;
		this.clientSessionId = clientSessionId;
		this.listener = listener;
		this.paymentRequest = paymentRequest;
	}

	/**
	 * Helper for encrypting the {@link PaymentRequest}.
	 *
	 * @param paymentRequest the {@link PaymentRequest} that will be encrypted
	 * @param clientSessionId the sessionId that is used to communicate with the Online Payments gateway
	 * @param metaData the metadata which is sent to the Online Payments gateway
	 * @param listener {@link OnPaymentRequestPreparedListener} that will be called when encryption is completed
	 *
	 * @deprecated Use {@link SessionEncryptionHelper#SessionEncryptionHelper(PaymentRequest, String, Map, PaymentRequestPreparedListener)} instead
	 */
	@Deprecated
	public SessionEncryptionHelper(PaymentRequest paymentRequest, String clientSessionId, Map<String, String> metaData, OnPaymentRequestPreparedListener listener) {

		if (paymentRequest == null ) {
			throw new InvalidParameterException("Error creating SessionEncryptionHelper, paymentRequest may not be null");
		}
		if (clientSessionId == null ) {
			throw new InvalidParameterException("Error creating SessionEncryptionHelper, clientSessionId may not be null");
		}
		if (listener == null ) {
			throw new InvalidParameterException("Error creating SessionEncryptionHelper, listener may not be null");
		}
		if (metaData == null) {
			throw new InvalidParameterException("Error creating SessionEncryptionHelper, metaData may not be null");
		}

		this.clientSessionId = clientSessionId;
		this.listener = listener;
		this.paymentRequest = paymentRequest;
		this.metaData = metaData;
	}

	/**
	 * Helper for encrypting the {@link PaymentRequest}.
	 *
	 * @param paymentRequest the {@link PaymentRequest} that will be encrypted
	 * @param clientSessionId the sessionId that is used to communicate with the Online Payments gateway
	 * @param metaData the metadata which is sent to the Online Payments gateway
	 * @param listener {@link PaymentRequestPreparedListener} that will be called when encryption is completed
	 */
	public SessionEncryptionHelper(PaymentRequest paymentRequest, String clientSessionId, Map<String, String> metaData, PaymentRequestPreparedListener listener) {

		if (paymentRequest == null ) {
			throw new InvalidParameterException("Error creating SessionEncryptionHelper, paymentRequest may not be null");
		}
		if (clientSessionId == null ) {
			throw new InvalidParameterException("Error creating SessionEncryptionHelper, clientSessionId may not be null");
		}
		if (listener == null ) {
			throw new InvalidParameterException("Error creating SessionEncryptionHelper, listener may not be null");
		}
		if (metaData == null) {
			throw new InvalidParameterException("Error creating SessionEncryptionHelper, metaData may not be null");
		}

		this.clientSessionId = clientSessionId;
		this.paymentRequestPreparedListener = listener;
		this.paymentRequest = paymentRequest;
		this.metaData = metaData;
	}


	/**
	 * Listener for loaded public key as {@link PublicKeyResponse} from the Online Payments gateway.
	 */
	@Override
	public void onPublicKeyLoaded(PublicKeyResponse response) {
	     onPublicKeyReceived(response);
	}

	@Override
	public void onPublicKeyError(ErrorResponse error) {
		Log.w("PublicKey", error.message + error.apiError);
		if (paymentRequestPreparedListener != null) {
			paymentRequestPreparedListener.onFailure(new EncryptDataException("Error while retrieving public key: " + error.message));
		}
	}

	/**
	 * Notify listeners for encryption result.
	 */
	@Override
	public void onEncryptDataComplete(String encryptedData) {
		notifyOldListener(encryptedData);
		notifyListener(encryptedData);
	}

	private void notifyListener(@Nullable String encryptedData) {
		if (paymentRequestPreparedListener != null) {
			if (encryptedData == null) {
				paymentRequestPreparedListener.onFailure(new EncryptDataException("Encryption result is null"));
				return;
			}

			// Call the OnPaymentRequestPrepared listener with the new PreparedPaymentRequest()
			if (metaData != null && !metaData.isEmpty()) {
				paymentRequestPreparedListener.onPaymentRequestPrepared(new PreparedPaymentRequest(encryptedData, Util.getBase64EncodedMetadata(metaData)));
			} else {
				paymentRequestPreparedListener.onPaymentRequestPrepared(new PreparedPaymentRequest(encryptedData, Util.getBase64EncodedMetadata(context)));
			}
		}
	}

	/**
	 * Temporary function to handle deprecated listener
	 * @param encryptedData String representing the encrypted fields for the payment request
	 */
	@Deprecated
	private void notifyOldListener(String encryptedData) {
		if (listener != null) {
			// Call the OnPaymentRequestPrepared listener with the new PreparedPaymentRequest()
			if (metaData != null && !metaData.isEmpty()) {
				listener.onPaymentRequestPrepared(new PreparedPaymentRequest(encryptedData, Util.getBase64EncodedMetadata(metaData)));
			} else {
				listener.onPaymentRequestPrepared(new PreparedPaymentRequest(encryptedData, Util.getBase64EncodedMetadata(context)));
			}
		}

	}

	public void onPublicKeyReceived(@NonNull PublicKeyResponse response) {
		EncryptDataAsyncTask task = new EncryptDataAsyncTask(response, paymentRequest, clientSessionId, this);
		task.execute();
	}


	/**
     * Interface for OnPaymentRequestPreparedListener.
     * Is called from the {@link Session} when it has encrypted the given payment product fields and composed the {@link PreparedPaymentRequest} object with them.
	 *
	 * @deprecated Use {@link PaymentRequestPreparedListener} instead
     */
	@Deprecated
    public interface OnPaymentRequestPreparedListener {
        void onPaymentRequestPrepared(PreparedPaymentRequest preparedPaymentRequest);
    }
}

File: onlinepayments-sdk/src/main/kotlin/com/onlinepayments/sdk/client/android/session/SessionEncryptionHelper.kt
/*
 * Do not remove or alter the notices in this preamble.
 *
 * Copyright © 2025 Worldline and/or its affiliates.
 *
 * All rights reserved. License grant and user rights and obligations according to the applicable license agreement.
 *
 * Please contact Worldline for questions regarding license and user rights.
 */

package com.onlinepayments.sdk.client.android.session

import android.content.Context
import com.onlinepayments.sdk.client.android.configuration.Constants
import com.onlinepayments.sdk.client.android.encryption.EncryptData
import com.onlinepayments.sdk.client.android.encryption.Encryptor
import com.onlinepayments.sdk.client.android.exception.EncryptDataException
import com.onlinepayments.sdk.client.android.listener.PaymentRequestPreparedListener
import com.onlinepayments.sdk.client.android.model.CreditCardTokenRequest
import com.onlinepayments.sdk.client.android.model.PaymentRequest
import com.onlinepayments.sdk.client.android.model.PreparedPaymentRequest
import com.onlinepayments.sdk.client.android.model.PublicKeyResponse
import com.onlinepayments.sdk.client.android.model.paymentproduct.PaymentProductField
import com.onlinepayments.sdk.client.android.util.Util

/**
 * Helper for encrypting the [PaymentRequest].
 *
 * @param context [Context] used for reading device metadata which is sent to the Online Payments gateway
 * @param clientSessionId the sessionId that is used to communicate with the Online Payments gateway
 * @param metaData the metadata which is sent to the Online Payments gateway
 * @param paymentRequestPreparedListener [PaymentRequestPreparedListener] [Deprecated] that is called when encryption is completed
 */
internal sealed class SessionEncryptionHelper(
    val context: Context,
    val clientSessionId: String,
    val metaData: Map<String, String>,
    val paymentRequestPreparedListener: PaymentRequestPreparedListener? = null
) {
    protected fun getResponse(encryptedData: String): PreparedPaymentRequest {
        return if (metaData.isNotEmpty()) {
            PreparedPaymentRequest(encryptedData, Util.getBase64EncodedMetadata(metaData))
        } else {
            PreparedPaymentRequest(
                encryptedData,
                Util.getBase64EncodedMetadata(context, null, Constants.SDK_IDENTIFIER)
            )
        }
    }

    fun getPreparedRequest(publicKey: PublicKeyResponse): PreparedPaymentRequest {
        val encrypted = encrypt(publicKey, encryptData(publicKey))
        return getResponse(encrypted)
    }

    fun onPublicKeyReceived(publicKey: PublicKeyResponse) {
        try {
            val encrypted = encrypt(publicKey, encryptData(publicKey))
            paymentRequestPreparedListener?.onPaymentRequestPrepared(getResponse(encrypted))
        } catch (e: EncryptDataException) {
            paymentRequestPreparedListener?.onFailure(e)
        }
    }

    protected abstract fun encryptData(publicKey: PublicKeyResponse): EncryptData

    private fun encrypt(publicKey: PublicKeyResponse, data: EncryptData): String =
        Encryptor(publicKey).encrypt(data)

    protected fun formattedPaymentValues(req: PaymentRequest): Map<String, String> {
        val formatted = mutableMapOf<String, String>()
        for (field in req.paymentProduct!!.getPaymentProductFields()) {
            req.getValue(field.id)?.let { value ->
                val unmasked = req.getUnmaskedValue(field.id, value)
                field.type?.let { type ->
                    formatted[field.id] =
                        if (type == PaymentProductField.Type.NUMERICSTRING)
                            unmasked.replace("[^\\d.]".toRegex(), "")
                        else unmasked
                }
            }
        }
        return formatted
    }

    protected fun filterTokenRequestValues(req: CreditCardTokenRequest): Map<String, String> {
        return req.getValues()
            .filterValues { value ->
                when (value) {
                    null -> false
                    is String -> value.isNotBlank()
                    else -> true
                }
            }
            .mapValues { (_, value) -> value.toString() }
    }

    protected fun formattedUnmaskedValue(
        fieldType: PaymentProductField.Type,
        unmaskedValue: String
    ): String {
        return if (fieldType == PaymentProductField.Type.NUMERICSTRING) {
            unmaskedValue.replace("[^\\d.]".toRegex(), "")
        } else {
            unmaskedValue
        }
    }

    class Payment(
        context: Context,
        clientSessionId: String,
        metaData: Map<String, String>,
        paymentRequestPreparedListener: PaymentRequestPreparedListener? = null,
        private val paymentRequest: PaymentRequest
    ) : SessionEncryptionHelper(context, clientSessionId, metaData, paymentRequestPreparedListener) {

        override fun encryptData(publicKey: PublicKeyResponse): EncryptData {
            val nonce = java.util.UUID.randomUUID().toString()
            val accountOnFileId = paymentRequest.accountOnFile?.id
            val productId = paymentRequest.paymentProduct?.getId()?.toInt()
            val values = formattedPaymentValues(paymentRequest)

            return EncryptData(
                accountOnFileId = accountOnFileId,
                clientSessionId = clientSessionId,
                nonce = nonce,
                paymentProductId = productId,
                tokenize = paymentRequest.tokenize,
                paymentValues = values
            )
        }
    }

    class TokenPayment(
        context: Context,
        clientSessionId: String,
        metaData: Map<String, String>,
        paymentRequestPreparedListener: PaymentRequestPreparedListener? = null,
        private val tokenRequest: CreditCardTokenRequest
    ) : SessionEncryptionHelper(context, clientSessionId, metaData, paymentRequestPreparedListener) {

        override fun encryptData(publicKey: PublicKeyResponse): EncryptData {
            val nonce = java.util.UUID.randomUUID().toString()
            val values = filterTokenRequestValues(tokenRequest)
            return EncryptData(
                accountOnFileId = null,
                clientSessionId = clientSessionId,
                nonce = nonce,
                paymentProductId = tokenRequest.paymentProductId,
                tokenize = false,
                paymentValues = values
            )
        }
    }
}
